xbmc/video/tags/VideoTagLoaderPlugin.cpp

```diff
@@ -0,0 +1,77 @@
+/*
+ *      Copyright (C) 2005-2017 Team Kodi
+ *      http://kodi.tv
+ *
+ *  This Program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2, or (at your option)
+ *  any later version.
+ *
+ *  This Program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with Kodi; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "VideoTagLoaderPlugin.h"
+#include "FileItem.h"
+#include "URL.h"
+#include "filesystem/PluginDirectory.h"
+
+using namespace XFILE;
+
+CVideoTagLoaderPlugin::CVideoTagLoaderPlugin(const CFileItem& item, bool forceRefresh)
+  : IVideoInfoTagLoader(item, nullptr, false), m_force_refresh(forceRefresh)
+{
+  if (forceRefresh)
+    return;
+  // Preserve CFileItem video info and art to avoid info loss between creating VideoInfoTagLoaderFactory and calling Load()
+  if (m_item.HasVideoInfoTag())
+    m_tag.reset(new CVideoInfoTag(*m_item.GetVideoInfoTag()));
+  auto& art = item.GetArt();
+  if (!art.empty())
+    m_art.reset(new CGUIListItem::ArtMap(art));
+}
+
+bool CVideoTagLoaderPlugin::HasInfo() const
+{
+  return m_tag || m_force_refresh;
+}
+
+CInfoScanner::INFO_TYPE CVideoTagLoaderPlugin::Load(CVideoInfoTag& tag, bool, std::vector<EmbeddedArt>*)
+{
+  if (m_force_refresh)
+  {
+    // In case of force refresh call our plugin with option "kodi_action=refresh_info"
+    // Plugin must do all refreshing work at specified path and return directory containing one ListItem with video tag and art
+    // We cannot obtain all info from setResolvedUrl, because CPluginDirectory::GetPluginResult doesn't copy full art
+    CURL url(m_item.GetPath());
+    url.SetOption("kodi_action", "refresh_info");
```

Comment:
What is this for? How is it different from a 'normal' listing? It smells leaking internal details of scraping into plugins..

Reply:
If a user press "Refresh" button in video info dialog we have to inform plugin that it should refresh specified item instead of list/play.

Example 1:
plugin://plugin.video.testlib/tvshows/?sid=123 - plugin should list tv show 123 contents (returns episodes)
plugin://plugin.video.testlib/tvshows/?sid=123&kodi_action=refresh_info - plugin should refresh tv show 123 information and return tv show folder, not episodes

Example 2:
plugin://plugin.video.testlib/tvshows/?mid=123 - plugin should play movie 123
plugin://plugin.video.testlib/tvshows/?mid=123&kodi_action=refresh_info - plugin should refresh movie 123 information and return movie details instead of play it

Reply:
And what if the plugin doesn't implement and respect this policy?

Reply:
Here's the thing: fetching metadata from the directory itself is different from when you have check a remote source to find out if these are changes. 'Refresh' is just a workaround for the latter. When you always get updated metadata on listing the parent directory, it doesn't make any sense to manually refresh it.

Reply:
Scanning and refreshing is a two different things. Scanning only adds new elements to media library. Refreshing updates current item information e.g. video info tag and art. Currently the scanner do not update information in existing items, moreover it just skips them in case of equal hashes (hash is calculated with subitems path, size and date).

> And what if the plugin doesn't implement and respect this policy?

Then "Refresh" button will not work. This is a new functionality and it's normal. As I've written above, if this changes will be approved, later we can implement some protection, so plugins without media library scanning support will not be scanned and refreshed even if they are used as video source.

Reply:
Well in that case the hashing is just plain wrong as it doesn't reflect the content. The point is this whole manual refreshing thing is legacy and something we want gone in the future, so hard-coding that in the plugin api is not a very good idea.

Reply:
If I understand correctly, there were no plans for current media library concept to reflect content changes during scanning. Only add new elements to library, so hashing uses only path and file information.
```
if (m_database.HasMovieInfo(pItem->GetPath()))
  return INFO_HAVE_ALREADY;
```
I'm making this changes for current media library functionality, and it will be a complete solution only as combination with such plugin related things (actually only "kodi_action=refresh_info" is desirable to make "Refresh" button work).

You have some plans for changing media library concepts, there are Montellese's project and others. I understand that in the future a lot of things can change, but when this future will come? Should we all wait for the "true" solution, that will be implemented maybe in Kodi 19 or maybe in Kodi 20, and stay with broken functionality for years? If in the future someone implement new better solution, then we all will use this new solution, but currently we have Kodi 18 and we can already use it with some small improvements like this.

Reply:
If you keep adding workarounds on top of workarounds it be 40 at least I can assure you that.

Reply:
We have different goals. You want to make a proper solution by reworking media library and changing its concepts and principles. You are absolutely right, but when will it be implemented? My goal is to keep current media library functionality unchanged and just fix something that do not work (scanning plugin sources in conjuction with metadata.local scraper). I was not going to change anything global.

Until there is no better solution, may we use this?